src/brevitas/graph/calibrate.py

```diff
@@ -195,41 +216,46 @@ def correct_bias_hook(self, module, inp, name, parent_module):
             quant_mean = self.compute_mean(inp, transpose_dim)
             error = self.float_mean_map[name] - quant_mean
             self.update_correction(name, error)
-            if self.curr_iter == self.iterations - 1:
-                self.apply_correction(name, parent_module)
             del self.float_mean_map[name]
             inp_broadcast_shape = compute_channel_view_shape(inp, channel_dim=self.channel_dim(inp, parent_module))
             return inp + error.reshape(inp_broadcast_shape)
 
-    def collect_float_mean(self, model, *args, **kwargs):
+    def register_collect_float_mean_hook(self, model):
         for name, module in model.named_modules():
             if isinstance(module, self.layers):
                 hook_fn = partial(self.collect_float_mean_hook, name=name, parent_module=module)
                 hook = module.output_quant.register_forward_pre_hook(hook_fn)
                 self.collect_float_mean_hooks.append(hook)
-        model(*args, **kwargs)
-        for hook in self.collect_float_mean_hooks:
-            hook.remove()
-        self.collect_float_mean_hooks = []
 
-    def correct_bias(self, model, *args, **kwargs):
+    def register_correct_bias_hook(self, model):
         for name, module in model.named_modules():
             if isinstance(module, self.layers):
                 hook_fn = partial(self.correct_bias_hook, name=name, parent_module=module)
                 hook = module.output_quant.register_forward_pre_hook(hook_fn)
                 self.correct_bias_hooks.append(hook)
-        model(*args, **kwargs)
+
+    def float_mean_hooks_cleanup(self):
+        for hook in self.collect_float_mean_hooks:
+            hook.remove()
+        self.collect_float_mean_hooks = []
+
+    def correct_bias_hooks_cleanup(self):
         for hook in self.correct_bias_hooks:
             hook.remove()
         self.correct_bias_hooks = []
-        assert not self.float_mean_map
 
-    def apply(self, model, *model_args, **model_kwargs):
+    def global_forward_pre_hook(self, model, inp):
         self.disable_act_quantization(model, is_training=False)
         self.disable_param_quantization(model, is_training=False)
-        self.collect_float_mean(model, *model_args, **model_kwargs)
+        self.register_collect_float_mean_hook(model)
+        inp, = inp
+        model.forward(inp) # Required to avoid infinite recursion
```

Comment:
this way we are skipping any potential hooks that the user might have defined on the model. Can we instead wrap the model in a module and register the hook on that, and call the forward of the wrapper?

Reply:
We don't skip any hook because outside the user is still calling `model(inp)`, that executes forward_pre_hooks and forward_hooks.

We are just using a forward_pre_hook to execute a hidden forward without quantization. 
Then we re-enable quantization, the execution continues with the "original forward", then the forward_hooks would be called (and we have one to remove our hooks for the next iteration).

If there are other hooks, they would be called just as ours are. In this way, we are also guaranteeing that if the user set a forward_hook, they would see the original quantized model output.

Reply:
I will investigate more to be sure that everything works as expected, maybe add some tests if you have particular scenarios that we need to preserve.